readthedocs/doc_builder/director.py

```diff
@@ -335,6 +339,163 @@ def run_build_job(self, job):
         for command in commands:
             environment.run(*command.split(), escape_command=False, cwd=cwd)
 
+    def run_build_commands(self):
+        cwd = self.data.project.checkout_path(self.data.version.slug)
+        environment = self.vcs_environment
+        for command in self.data.config.build.commands:
+            environment.run(*command.split(), escape_command=False, cwd=cwd)
+
+        # Copy files to artifacts path so they are uploaded to S3
+        target = self.data.project.artifact_path(
+            version=self.data.version.slug,
+            type_="sphinx",
```

Comment:
I think we should introduce a new doctype, the doctype is used in several places to make some decisions (specially when generating links, foo/ vs /foo.html). It should also make it easier to filter projects/versions using this new feature.

Reply:
I'm not aware of how these `type_`s are used. If we use `generic` here, what other changes in logic do we need to do? Will it touch multiple places?

Reply:
The more prominent places where this used are

- version.documentation_type
   - https://github.com/readthedocs/readthedocs.org/blob/d3351d52c41507f4c5f1ba737f8ef0b684d6e642/readthedocs/api/v2/views/footer_views.py#L158-L158
   - https://github.com/readthedocs/readthedocs.org/blob/d3351d52c41507f4c5f1ba737f8ef0b684d6e642/readthedocs/proxito/views/serve.py#L353-L353
   - https://github.com/readthedocs/readthedocs.org/blob/a59150c674fbb3bdc426183c623fe1ac37e64ec0/readthedocs/search/serializers.py#L137-L137

- version.is_sphinx_type

I have also wanted to get rid of using the doctype from some places (footer and search)

Reply:
I'll try to add a `GENERIC` constant and see what happens. After an initial grep of `DOCUMENTATION_CHOICES` and `DOCTYPE_CHOICES` I didn't find it used in many places.

Also, grepping for `SPHINX` I didn't find anything that could conflict with `GENERIC`. The code pointed by @stsewd makes sense but I think it will be fine.

Besides, as @ericholscher thought it seems that usually, the default is `SPHINX` and the special cases are for `SPHINX_HTMLDIR`. So, I think it could work using a new doctype 👍🏼

Reply:
I added a `GENERIC` type and I didn't find too many problems. Everything seems to continue working fine. I'm not sure if we will find some issues in the future because of this, but at least nothing obvious at this point.

As I said, I'm not really aware of how the doctype is used in the code, but I checked all the files I was pointed to and I didn't find anything that called my attention and I thought it could break at least.

Actually, while working on this I found a bug. We weren't updating the `Version.documentation_type` based on the `doctype` used in the config file --so, all the `Version.documentation_type` are defaulting to the `Project.documentation_type`. See this commit for the fix: [`99efb00` (#9150)](https://github.com/readthedocs/readthedocs.org/pull/9150/commits/99efb00bad5138d7d609b9474840c416dce3f560)

Let me know if there is something else here I should care about.